<a href="https://colab.research.google.com/github/zingarelli/alura-challenges-data-science-2022/blob/main/Alura_Challenges_%7C_Data_Science_2022_Semana_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Carregamento da Base de Dados

A base de dados será carregada diretamente do link do GitHub disponibilizado pela empresa. A base é um arquivo JSON.

O dicionário de dados, com as informações sobre cada coluna da base, pode ser visto em https://github.com/sthemonica/alura-voz/blob/main/dicionario.md

In [1]:
import pandas as pd

df = pd.read_json(
    path_or_buf='https://raw.githubusercontent.com/sthemonica/alura-voz/main/Dados/Telco-Customer-Churn.json',
)

df.head()

customerID Churn                                           customer  \
0  0002-ORFBO    No  {'gender': 'Female', 'SeniorCitizen': 0, 'Part...   
1  0003-MKNFE    No  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
2  0004-TLHLJ   Yes  {'gender': 'Male', 'SeniorCitizen': 0, 'Partne...   
3  0011-IGKFF   Yes  {'gender': 'Male', 'SeniorCitizen': 1, 'Partne...   
4  0013-EXCHZ   Yes  {'gender': 'Female', 'SeniorCitizen': 1, 'Part...   

                                             phone  \
0   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
1  {'PhoneService': 'Yes', 'MultipleLines': 'Yes'}   
2   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
3   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   
4   {'PhoneService': 'Yes', 'MultipleLines': 'No'}   

                                            internet  \
0  {'InternetService': 'DSL', 'OnlineSecurity': '...   
1  {'InternetService': 'DSL', 'OnlineSecurity': '...   
2  {'InternetService': 'Fiber optic', 'OnlineSecu...   
3  {'InternetService': 'Fiber optic', 'OnlineSecu...   
4  {'InternetService': 'Fiber optic', 'OnlineSecu...   

                                             account  
0  {'Contract': 'One year', 'PaperlessBilling': '...  
1  {'Contract': 'Month-to-month', 'PaperlessBilli...  
2  {'Contract': 'Month-to-month', 'PaperlessBilli...  
3  {'Contract': 'Month-to-month', 'PaperlessBilli...  
4  {'Contract': 'Month-to-month', 'PaperlessBilli...

Os dados já vem com uma certa divisão, porém, é possível ver que há mais níveis de profundidade em quase todas as colunas. Então, utilizaremos o json_normalize nestas colunas para aprofundar na tabulação dos dados

In [2]:
# coluna customer
customer = pd.json_normalize(data = df.customer)
customer.head()

gender  SeniorCitizen Partner Dependents  tenure
0  Female              0     Yes        Yes       9
1    Male              0      No         No       9
2    Male              0      No         No       4
3    Male              1     Yes         No      13
4  Female              1     Yes         No       3

In [3]:
# coluna phone
phone = pd.json_normalize(data = df.phone)
phone.head()

PhoneService MultipleLines
0          Yes            No
1          Yes           Yes
2          Yes            No
3          Yes            No
4          Yes            No

In [4]:
# coluna internet
internet = pd.json_normalize(data = df.internet)
internet.head()

InternetService OnlineSecurity OnlineBackup DeviceProtection TechSupport  \
0             DSL             No          Yes               No         Yes   
1             DSL             No           No               No          No   
2     Fiber optic             No           No              Yes          No   
3     Fiber optic             No          Yes              Yes          No   
4     Fiber optic             No           No               No         Yes   

  StreamingTV StreamingMovies  
0         Yes              No  
1          No             Yes  
2          No              No  
3         Yes             Yes  
4         Yes              No

In [5]:
# coluna account
account = pd.json_normalize(data = df.account)
account.head()

Contract PaperlessBilling     PaymentMethod  Charges.Monthly  \
0        One year              Yes      Mailed check             65.6   
1  Month-to-month               No      Mailed check             59.9   
2  Month-to-month              Yes  Electronic check             73.9   
3  Month-to-month              Yes  Electronic check             98.0   
4  Month-to-month              Yes      Mailed check             83.9   

  Charges.Total  
0         593.3  
1         542.4  
2        280.85  
3       1237.85  
4         267.4

Concatenando todas as colunas em um único DataFrame

In [6]:
data = pd.concat([df[['customerID',	'Churn']], customer, phone, internet, account], axis = 1)

data.head()

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
0  0002-ORFBO    No  Female              0     Yes        Yes       9   
1  0003-MKNFE    No    Male              0      No         No       9   
2  0004-TLHLJ   Yes    Male              0      No         No       4   
3  0011-IGKFF   Yes    Male              1     Yes         No      13   
4  0013-EXCHZ   Yes  Female              1     Yes         No       3   

  PhoneService MultipleLines InternetService  ... OnlineBackup  \
0          Yes            No             DSL  ...          Yes   
1          Yes           Yes             DSL  ...           No   
2          Yes            No     Fiber optic  ...           No   
3          Yes            No     Fiber optic  ...          Yes   
4          Yes            No     Fiber optic  ...           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No         Yes         Yes              No        One year   
1               No          No          No             Yes  Month-to-month   
2              Yes          No          No              No  Month-to-month   
3              Yes          No         Yes             Yes  Month-to-month   
4               No         Yes         Yes              No  Month-to-month   

  PaperlessBilling     PaymentMethod Charges.Monthly  Charges.Total  
0              Yes      Mailed check            65.6          593.3  
1               No      Mailed check            59.9          542.4  
2              Yes  Electronic check            73.9         280.85  
3              Yes  Electronic check            98.0        1237.85  
4              Yes      Mailed check            83.9          267.4  

[5 rows x 21 columns]

# Entendendo o DataFrame

Vamos agora ter uma noção inicial da base de dados, quais informações ela possui e os tipos de dados em cada coluna.

In [7]:
# quantidade de registros e colunas
print(f'A base de dados possui {data.shape[0]} registros e {data.shape[1]} colunas.')

A base de dados possui 7267 registros e 21 colunas.


In [8]:
# informações sobre as colunas
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7267 entries, 0 to 7266
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7267 non-null   object 
 1   Churn             7267 non-null   object 
 2   gender            7267 non-null   object 
 3   SeniorCitizen     7267 non-null   int64  
 4   Partner           7267 non-null   object 
 5   Dependents        7267 non-null   object 
 6   tenure            7267 non-null   int64  
 7   PhoneService      7267 non-null   object 
 8   MultipleLines     7267 non-null   object 
 9   InternetService   7267 non-null   object 
 10  OnlineSecurity    7267 non-null   object 
 11  OnlineBackup      7267 non-null   object 
 12  DeviceProtection  7267 non-null   object 
 13  TechSupport       7267 non-null   object 
 14  StreamingTV       7267 non-null   object 
 15  StreamingMovies   7267 non-null   object 
 16  Contract          7267 non-null   object 


Temos 7267 registros. Nenhuma coluna possui dados nulos. A maioria das colunas contém dados categóricos, com a exceção sendo em:


*   Charges.Monthly, que é float64;
*   SeniorCitizen e tenure, marcados como int64.

Baseado no dicionário de dados, o tipo está consistente em quase todas as colunas, exceto em `Charges.Total`. Ela contém o valor total gasto pelo cliente, o que se supõe ser um valor numérico, igual à coluna `Charges.Montly`. No entanto, seu dtype está marcado como object. Talvez alguns valores inválidos foram registrados nela e por isso ela foi classificada como object. Vamos fazer a conversão para float64 e marcar os erros de conversão como NaN.



## Resolvendo a inconsistência em `Charges.Total`

In [9]:
# verificando algumas estatísticas da coluna Charges.Total
data['Charges.Total'].describe()

count     7267
unique    6531
top           
freq        11
Name: Charges.Total, dtype: object

In [10]:
# convertendo o conteúdo da coluna para float. O coerce marca os erros de conversão como sendo NaN
charges_total = pd.to_numeric(data['Charges.Total'], errors = 'coerce')

charges_total.dtype

dtype('float64')

In [11]:
# verificando algumas estatísticas dos dados convertidos
charges_total.describe()

count    7256.000000
mean     2280.634213
std      2268.632997
min        18.800000
25%       400.225000
50%      1391.000000
75%      3785.300000
max      8684.800000
Name: Charges.Total, dtype: float64

In [12]:
# verificando quantos registros não eram numéricos e foram marcados como NaN
charges_total.isnull().sum()

11

Na variável que foi criada, `charges_total`, vemos que 11 registros possuíam conteúdo não-numérico. Como são poucos registros, vamos olhar o conteúdo original deles na base.

In [13]:
# indices que contém NaN
nan_index = charges_total[charges_total.isnull()].index

# selecionando os registros na base para ver o conteúdo original
data.iloc[nan_index]['Charges.Total'].values

array([' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' ', ' '],
      dtype=object)

Como podemos observar, nos 11 registros há uma string com espaço em branco `(' ')`.

`Charges.Total` contabiliza o total gasto pelo cliente desde que assinou seu plano. Uma alternativa para sanar a inconsistência nestes 11 registros é fazer a multiplicação entre as colunas `tenure` e `Charges.Monthly`, obtendo uma aproximação do valor total gasto (talvez não seja totalmente acurado, pois não é possível saber se o valor mensal é fixo ou se pode variar). 



In [14]:
# Verificando as colunas tenure e Charges.Montly dos 11 registros
data.iloc[nan_index][['tenure', 'Charges.Monthly']]

tenure  Charges.Monthly
975        0            56.05
1775       0            20.00
1955       0            61.90
2075       0            19.70
2232       0            20.25
2308       0            25.35
2930       0            73.35
3134       0            25.75
3203       0            52.55
4169       0            80.85
5599       0            19.85

É possível notar que o valor de tenure nestes 11 registros é zero. Ou seja, são novos clientes, que ainda não completaram 1 mês de serviço. Nesse caso, podemos então atribuir à coluna `Charges.Total` o valor que está na coluna `Charges.Monthly`

In [15]:
# preenchendo os valores vazios com os valores mensais na variável charges_total
charges_total.fillna(data['Charges.Monthly'], inplace=True)

# verificando se há algum registro nulo
charges_total.isnull().sum()

0

In [16]:
# substituindo os valores na base pelos valores da variável charges_total, usando o loc 
data['Charges.Total'] = charges_total

# conferindo a alteração
data.iloc[nan_index][['Charges.Monthly', 'Charges.Total']]

Charges.Monthly  Charges.Total
975             56.05          56.05
1775            20.00          20.00
1955            61.90          61.90
2075            19.70          19.70
2232            20.25          20.25
2308            25.35          25.35
2930            73.35          73.35
3134            25.75          25.75
3203            52.55          52.55
4169            80.85          80.85
5599            19.85          19.85

In [17]:
# conferindo o dtype da coluna 
data['Charges.Total'].dtype

dtype('float64')

## Analisando e resolvendo inconsistências nas variáveis categóricas

Vamos agora verificar os valores contidos em cada variável categórica e se estão consistentes

In [18]:
# selecionando as colunas categóricas
categorical_cols = data.select_dtypes(include=object).columns

# verificando os valores únicos de cada coluna categórica
for col in categorical_cols:
  print(f'Coluna: {col}\nValores únicos: {data[col].unique()}\n')

Coluna: customerID
Valores únicos: ['0002-ORFBO' '0003-MKNFE' '0004-TLHLJ' ... '9992-UJOEL' '9993-LHIEB'
 '9995-HOTOH']

Coluna: Churn
Valores únicos: ['No' 'Yes' '']

Coluna: gender
Valores únicos: ['Female' 'Male']

Coluna: Partner
Valores únicos: ['Yes' 'No']

Coluna: Dependents
Valores únicos: ['Yes' 'No']

Coluna: PhoneService
Valores únicos: ['Yes' 'No']

Coluna: MultipleLines
Valores únicos: ['No' 'Yes' 'No phone service']

Coluna: InternetService
Valores únicos: ['DSL' 'Fiber optic' 'No']

Coluna: OnlineSecurity
Valores únicos: ['No' 'Yes' 'No internet service']

Coluna: OnlineBackup
Valores únicos: ['Yes' 'No' 'No internet service']

Coluna: DeviceProtection
Valores únicos: ['No' 'Yes' 'No internet service']

Coluna: TechSupport
Valores únicos: ['Yes' 'No' 'No internet service']

Coluna: StreamingTV
Valores únicos: ['Yes' 'No' 'No internet service']

Coluna: StreamingMovies
Valores únicos: ['No' 'Yes' 'No internet service']

Coluna: Contract
Valores únicos: ['One year' 'Month-

É possível observar que na coluna `Churn` há valores vazios `''`. Precisamos analisar melhor os registros com esse valor vazio.

Também podemos observar que várias colunas possuem o valor `'No internet service'` e uma delas possui o valor `'No phone service'`. Será que poderíamos substituir esses valores pela resposta `'No'`, ficando assim somente com `Yes` e `No`?

### Resolvendo a inconsistência em `Churn`

In [19]:
print(f'Total de registros: {data.shape[0]}')
print(f'Registros sem valor na coluna Churn: {data[data.Churn == ""].shape[0]}')
print(f'Proporção: {data[data.Churn == ""].shape[0] / data.shape[0]}')

Total de registros: 7267
Registros sem valor na coluna Churn: 224
Proporção: 0.030824274115866246


Temos que 224 registros não possuem nada na coluna `Churn`, o que dá 3% do total de registros na base. Será que removê-los afetaria as análises de maneira significante?

In [20]:
# olhando o primeiro registro com valor vazio em Churn
data[data.Churn == ""].head(2)

customerID Churn  gender  SeniorCitizen Partner Dependents  tenure  \
30  0047-ZHDTW        Female              0      No         No      11   
75  0120-YZLQA          Male              0      No         No      71   

   PhoneService MultipleLines InternetService  ...         OnlineBackup  \
30          Yes           Yes     Fiber optic  ...                   No   
75          Yes            No              No  ...  No internet service   

       DeviceProtection          TechSupport          StreamingTV  \
30                   No                   No                   No   
75  No internet service  No internet service  No internet service   

        StreamingMovies        Contract PaperlessBilling  \
30                   No  Month-to-month              Yes   
75  No internet service        Two year              Yes   

                PaymentMethod Charges.Monthly  Charges.Total  
30  Bank transfer (automatic)            79.0          929.3  
75    Credit card (automatic)            19.9         1355.1  

[2 rows x 21 columns]

##TO DO
- Decidir se removo ou não colunas com Churn vazio
- o que fazer com no internet/phone service (talvez deixar para depois, pois não se trata de uma inconsistência?)

# Análise Exploratória

Hora de verificar algumas estatísticas dos dados, bem como analisar e corrigir inconsistências.